`gstreamermm` fails to build

### Describe the bug

`gstreamermm` fails to build. Seems to be [due to GCC 11 upgrade](https://gitlab.gnome.org/GNOME/gstreamermm/-/issues/11). A [fix has been merged upstream](https://gitlab.gnome.org/GNOME/gstreamermm/-/merge_requests/4). That said, the project is mostly abandoned. The last release was 4 years ago, and they recommend building from a release since building from source is much harder. I can confirm, as I am unable to make it build from source (but then again for me autotools is like discovering ancient tablets in the sand with hieroglyphs on them).

There are only a few packages that depend on `gstreamermm`. If it's too much hassle to get working may just want to retire these.

- [obs-gstreamer](https://search.nixos.org/packages?channel=unstable&from=0&size=50&sort=relevance&type=packages&query=obs-gstreamer)
- [subtitleeditor](https://search.nixos.org/packages?channel=unstable&from=0&size=50&sort=relevance&type=packages&query=subtitleeditor)
- [mapmap](https://search.nixos.org/packages?channel=unstable&from=0&size=50&sort=relevance&type=packages&query=mapmap)

### Steps To Reproduce

1. `nix-build -A gst_all_1.gstreamermm`

### Expected behavior

Build works.

### Additional context

End of log from building on current `master`:

```
make[2]: Leaving directory '/build/gstreamermm-1.10.0/gstreamer/gstreamermm'
Making all in tests
make[2]: Entering directory '/build/gstreamermm-1.10.0/tests'
make[2]: Nothing to be done for 'all'.
make[2]: Leaving directory '/build/gstreamermm-1.10.0/tests'
Making all in examples
make[2]: Entering directory '/build/gstreamermm-1.10.0/examples'
  CXX      audio_video_muxer/main.o
  CXX      all_media_player/main.o
  CXX      dynamic_changing_source/main.o
In file included from /nix/store/ajl4sxhfgqzch5cnwq9rlsk19wxkq13n-glib-2.72.0-dev/include/glib-2.0/glib/gthread.h:32,
                 from /nix/store/ajl4sxhfgqzch5cnwq9rlsk19wxkq13n-glib-2.72.0-dev/include/glib-2.0/glib/gasyncqueue.h:32,
                 from /nix/store/ajl4sxhfgqzch5cnwq9rlsk19wxkq13n-glib-2.72.0-dev/include/glib-2.0/glib.h:32,
                 from /nix/store/x5balm62kwx33v3psj33q558z5aclg6b-glibmm-2.66.2-dev/include/glibmm-2.4/glibmm/threads.h:30,
                 from ../gstreamer/gstreamermm/check.h:11,
                 from ../gstreamer/gstreamermm.h:63,
                 from audio_video_muxer/main.cc:8:
../gstreamer/gstreamermm/register.h: In function 'GType Gst::register_mm_type(const gchar*)':
/nix/store/ajl4sxhfgqzch5cnwq9rlsk19wxkq13n-glib-2.72.0-dev/include/glib-2.0/glib/gatomic.h:113:19: error: argument 2 of '__atomic_load' must not be a pointer to a 'volatile' type
  113 |     __atomic_load (gapg_temp_atomic, &gapg_temp_newval, __ATOMIC_SEQ_CST); \
      |     ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/nix/store/ajl4sxhfgqzch5cnwq9rlsk19wxkq13n-glib-2.72.0-dev/include/glib-2.0/glib/gthread.h:260:7: note: in expansion of macro 'g_atomic_pointer_get'
  260 |     (!g_atomic_pointer_get (location) &&                             \
      |       ^~~~~~~~~~~~~~~~~~~~
../gstreamer/gstreamermm/register.h:114:9: note: in expansion of macro 'g_once_init_enter'
  114 |     if (g_once_init_enter (&gonce_data)) {
      |         ^~~~~~~~~~~~~~~~~
In file included from /nix/store/ajl4sxhfgqzch5cnwq9rlsk19wxkq13n-glib-2.72.0-dev/include/glib-2.0/glib/gthread.h:32,
                 from /nix/store/ajl4sxhfgqzch5cnwq9rlsk19wxkq13n-glib-2.72.0-dev/include/glib-2.0/glib/gasyncqueue.h:32,
                 from /nix/store/ajl4sxhfgqzch5cnwq9rlsk19wxkq13n-glib-2.72.0-dev/include/glib-2.0/glib.h:32,
                 from /nix/store/x5balm62kwx33v3psj33q558z5aclg6b-glibmm-2.66.2-dev/include/glibmm-2.4/glibmm/threads.h:30,
                 from ../gstreamer/gstreamermm/check.h:11,
                 from ../gstreamer/gstreamermm.h:63,
                 from all_media_player/main.cc:8:
../gstreamer/gstreamermm/register.h: In function 'GType Gst::register_mm_type(const gchar*)':
/nix/store/ajl4sxhfgqzch5cnwq9rlsk19wxkq13n-glib-2.72.0-dev/include/glib-2.0/glib/gatomic.h:113:19: error: argument 2 of '__atomic_load' must not be a pointer to a 'volatile' type
  113 |     __atomic_load (gapg_temp_atomic, &gapg_temp_newval, __ATOMIC_SEQ_CST); \
      |     ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/nix/store/ajl4sxhfgqzch5cnwq9rlsk19wxkq13n-glib-2.72.0-dev/include/glib-2.0/glib/gthread.h:260:7: note: in expansion of macro 'g_atomic_pointer_get'
  260 |     (!g_atomic_pointer_get (location) &&                             \
      |       ^~~~~~~~~~~~~~~~~~~~
../gstreamer/gstreamermm/register.h:114:9: note: in expansion of macro 'g_once_init_enter'
  114 |     if (g_once_init_enter (&gonce_data)) {
      |         ^~~~~~~~~~~~~~~~~
make[2]: *** [Makefile:829: audio_video_muxer/main.o] Error 1
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [Makefile:829: all_media_player/main.o] Error 1
In file included from /nix/store/ajl4sxhfgqzch5cnwq9rlsk19wxkq13n-glib-2.72.0-dev/include/glib-2.0/glib/gthread.h:32,
                 from /nix/store/ajl4sxhfgqzch5cnwq9rlsk19wxkq13n-glib-2.72.0-dev/include/glib-2.0/glib/gasyncqueue.h:32,
                 from /nix/store/ajl4sxhfgqzch5cnwq9rlsk19wxkq13n-glib-2.72.0-dev/include/glib-2.0/glib.h:32,
                 from /nix/store/x5balm62kwx33v3psj33q558z5aclg6b-glibmm-2.66.2-dev/include/glibmm-2.4/glibmm/threads.h:30,
                 from ../gstreamer/gstreamermm/check.h:11,
                 from ../gstreamer/gstreamermm.h:63,
                 from dynamic_changing_source/main.cc:7:
../gstreamer/gstreamermm/register.h: In function 'GType Gst::register_mm_type(const gchar*)':
/nix/store/ajl4sxhfgqzch5cnwq9rlsk19wxkq13n-glib-2.72.0-dev/include/glib-2.0/glib/gatomic.h:113:19: error: argument 2 of '__atomic_load' must not be a pointer to a 'volatile' type
  113 |     __atomic_load (gapg_temp_atomic, &gapg_temp_newval, __ATOMIC_SEQ_CST); \
      |     ~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/nix/store/ajl4sxhfgqzch5cnwq9rlsk19wxkq13n-glib-2.72.0-dev/include/glib-2.0/glib/gthread.h:260:7: note: in expansion of macro 'g_atomic_pointer_get'
  260 |     (!g_atomic_pointer_get (location) &&                             \
      |       ^~~~~~~~~~~~~~~~~~~~
../gstreamer/gstreamermm/register.h:114:9: note: in expansion of macro 'g_once_init_enter'
  114 |     if (g_once_init_enter (&gonce_data)) {
      |         ^~~~~~~~~~~~~~~~~
make[2]: *** [Makefile:829: dynamic_changing_source/main.o] Error 1
make[2]: Leaving directory '/build/gstreamermm-1.10.0/examples'
make[1]: *** [Makefile:607: all-recursive] Error 1
make[1]: Leaving directory '/build/gstreamermm-1.10.0'
make: *** [Makefile:461: all] Error 2
error: builder for '/nix/store/h3xignaxqd14g4aniv3ajfx7mcihw3l3-gstreamermm-1.10.0.drv' failed with exit code 2;
```

End of log trying to build from source:

```
make  all-recursive
make[1]: Entering directory '/home/austin/Documents/gstreamermm'
Making all in tools
make[2]: Entering directory '/home/austin/Documents/gstreamermm/tools'
  CXX      extra_defs_gen/generate_defs_gst.o
  CXXLD    extra_defs_gen/generate_defs_gst
  CXX      extra_defs_gen/generate_plugin_module_defs_gst.o
  CXXLD    extra_defs_gen/generate_plugin_module_defs_gst
  CXX      extra_defs_gen/generate_plugin_gmmproc_file.o
extra_defs_gen/generate_plugin_gmmproc_file.cc:386:13: warning: Deprecated pre-processor symbol: replace with "(g_array_get_type ())"
  386 |                                          returnGType != G_TYPE_VALUE_ARRAY &&
      |             ^                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  CXXLD    extra_defs_gen/generate_plugin_gmmproc_file
make[2]: Leaving directory '/home/austin/Documents/gstreamermm/tools'
Making all in gstreamer/gstreamermm
make[2]: Entering directory '/home/austin/Documents/gstreamermm/gstreamer/gstreamermm'
make[2]: *** No rule to make target 'allocator.cc', needed by 'allocator.lo'.  Stop.
make[2]: Leaving directory '/home/austin/Documents/gstreamermm/gstreamer/gstreamermm'
make[1]: *** [Makefile:610: all-recursive] Error 1
make[1]: Leaving directory '/home/austin/Documents/gstreamermm'
make: *** [Makefile:464: all] Error 2
```


### Notify maintainers

@alesya-h @plcplc @erictapen @romildo 

### Metadata

```yaml
 - system: `"x86_64-linux"`
 - host os: `Linux 5.17.3, NixOS, 22.05 (Quokka), 22.05pre372012.bc41b01dd7a`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.8.0pre20220411_f7276bc`
 - channels(root): `"home-manager, nixos"`
 - channels(austin): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

--- a/pkgs/development/libraries/gstreamer/gstreamermm/default.nix
+++ b/pkgs/development/libraries/gstreamer/gstreamermm/default.nix
@@ -1,6 +1,7 @@
 { lib
 , stdenv
 , fetchurl
+, fetchpatch
 , pkg-config
 , file
 , glibmm
@@ -16,6 +17,14 @@ stdenv.mkDerivation rec {
     url = "mirror://gnome/sources/${pname}/${lib.versions.majorMinor version}/${pname}-${version}.tar.xz";
     sha256 = "0q4dx9sncqbwgpzma0zvj6zssc279yl80pn8irb95qypyyggwn5y";
   };
+
+  patches = [
+    (fetchpatch {
+      name = "${pname}-${version}.fix-build-against-glib-2.68.patch";
+      url = "https://gitlab.gnome.org/GNOME/gstreamermm/-/commit/37116547fb5f9066978e39b4cf9f79f2154ad425.patch";
+      sha256 = "sha256-YHtmOiOl4POwas3eWHsew3IyGK7Aq22MweBm3JPwyBM=";
+    })
+  ];
 
   outputs = [ "out" "dev" ];
 
@@ -45,11 +54,6 @@ stdenv.mkDerivation rec {
     license = licenses.lgpl21Plus;
     platforms = platforms.linux;
     maintainers = with maintainers; [ romildo ];
-    broken = true; # at 2022-09-24, logs at: https://termbin.com/n5hb
-      # ../gstreamer/gstreamermm/register.h: In function 'GType Gst::register_mm_type(const gchar*)':
-      # /nix/store/rwr0ly4girpnrq3sqvp3v2k40wb4hh3s-glib-2.72.3-dev/include/glib-2.0/glib/gatomic.h:113:19:
-      #   error: argument 2 of '__atomic_load' must not be a pointer to a 'volatile' type
-      #   113 |     __atomic_load (gapg_temp_atomic, &gapg_temp_newval, __ATOMIC_SEQ_CST);
   };
 
 }

--- a/pkgs/development/libraries/gstreamer/gstreamermm/default.nix
+++ b/pkgs/development/libraries/gstreamer/gstreamermm/default.nix
@@ -1,4 +1,13 @@
-{ lib, stdenv, fetchurl, pkg-config, file, glibmm, gst_all_1, gnome }:
+{ lib
+, stdenv
+, fetchurl
+, pkg-config
+, file
+, glibmm
+, gst_all_1
+, gnome
+}:
+
 stdenv.mkDerivation rec {
   pname = "gstreamermm";
   version = "1.10.0";
@@ -10,9 +19,15 @@ stdenv.mkDerivation rec {
 
   outputs = [ "out" "dev" ];
 
-  nativeBuildInputs = [ pkg-config file ];
+  nativeBuildInputs = [
+    pkg-config
+    file
+  ];
 
-  propagatedBuildInputs = [ glibmm gst_all_1.gst-plugins-base ];
+  propagatedBuildInputs = [
+    glibmm
+    gst_all_1.gst-plugins-base
+  ];
 
   enableParallelBuilding = true;
 

